Switch prompts for path from TDML generate to TDML execute in launch …

JeremyYao opened the pull request:
…config.

Add Default Test Case name when adding config with the blue Add Configuration... button

Closes #1347

### Description

Refer to #1347

### Wiki

- [ ] I have determined that no wiki updates are needed for these changes
   - yes, updates are needed
- [x] I have added following documentation for these changes
   - See https://github.com/apache/daffodil-vscode/pull/1403#issuecomment-3292625611 for more details


### How to test (sample test case)

#### Example Files

[test_images.zip](https://github.com/user-attachments/files/22125186/test_images.zip)

#### Validating generate still works

0. Delete launch.json if it exists
1. Open the launch config wizard. 
2. Ensure no path field shows for generate

<img width="465" height="205" alt="image" src="https://github.com/user-attachments/assets/a129c6b1-6f33-4151-9c1c-53bdebd8f5e1" />

3. Save the config. You should have something that looks like 

<img width="690" height="891" alt="image" src="https://github.com/user-attachments/assets/9511d574-a5a4-4611-9ce3-9a456f0194a4" />

4. Run `Wizard Config` and select a sample DFDL file and data file

6. Navigate to your OS's temp directory and verify generatedTDML.tdml was created and/or updated after the DFDL debugging session is over 

<img width="734" height="139" alt="image" src="https://github.com/user-attachments/assets/fa9c6fcf-a075-4398-b5c7-bbf2bd9b4c1f" />

7. Create the TDML file 

<img width="247" height="160" alt="image" src="https://github.com/user-attachments/assets/c4664a49-0432-4742-adc9-33a7ee94f602" />

I named the TDML file `1347.tdml`

#### Validate Execute

1. Go back to `Wizard Config`
2. Select TDML action to be execute
3. Type in the path to the TDML file previously created

<img width="478" height="218" alt="image" src="https://github.com/user-attachments/assets/75456249-ee4e-4d7f-9963-815dfd48ab12" />

4. Run the config

#### Execute with blank string in path

1. Create new config

<img width="425" height="405" alt="image" src="https://github.com/user-attachments/assets/6a904274-d16e-45bc-861b-aa2ad1bb83b7" />

2. Set execute to empty string path 

<img width="456" height="240" alt="image" src="https://github.com/user-attachments/assets/1ffb0b88-fe83-4317-afb2-2ee162dcf916" />

The following config item should be added 
```JSON
{
            "request": "launch",
            "type": "dfdl",
            "name": "Wizard Config 2",
            "schema": {
                "path": "${command:AskForSchemaName}",
                "rootName": null,
                "rootNamespace": null
            },
            "data": "${command:AskForDataName}",
            "debugServer": 4711,
            "infosetFormat": "xml",
            "infosetOutput": {
                "type": "file",
                "path": "${workspaceFolder}/target/infoset.xml"
            },
            "tdmlConfig": {
                "action": "execute",
                "path": "",
                "name": "Default Test Case"
            },
            "trace": true,
            "stopOnEntry": true,
            "useExistingServer": false,
            "openDataEditor": false,
            "openInfosetView": false,
            "openInfosetDiffView": false,
            "daffodilDebugClasspath": [],
            "dataEditor": {
                "port": 9000,
                "logging": {
                    "file": "${workspaceFolder}/dataEditor-${omegaEditPort}.log",
                    "level": "info"
                }
            },
            "dfdlDebugger": {
                "logging": {
                    "file": "${workspaceFolder}/daffodil-debugger.log",
                    "level": "INFO"
                }
            }
        }
```

3. Run this config
4. Validate prompt for .tdml file pops up
5. Select a .tdml file that has a test case name of `Default Test Case` (example `1347.tdml` created in prior step)
6. Verify debugging session runs smoothly

#### Execute with ${AskForValidatedTDMLPath}

Follow steps above, but use this launch config as mentioned in https://github.com/apache/daffodil-vscode/pull/1403#issuecomment-3271741062 instead

JeremyYao commented:
Please @ me when this gets merged in so I can update the wiki afterwards

CoverRyan commented on src/daffodilDebugger/debugger.ts:
```diff
@@ -78,7 +78,7 @@ async function getTDMLConfig(
     config.tdmlConfig.name =
       config.tdmlConfig.name || getDefaultTDMLTestCaseName()
-    config.tdmlConfig.path = config.tdmlConfig.path || getTmpTDMLFilePath()
+    config.tdmlConfig.path = getTmpTDMLFilePath()
```
This change is the same as the change made in [1388](https://github.com/apache/daffodil-vscode/pull/1388). The issue this resolves, [1347](https://github.com/apache/daffodil-vscode/issues/1347), seems to overlap with [1376](https://github.com/apache/daffodil-vscode/issues/1376) and the changes in 1388 in general.

JeremyYao replied:
Added `Closes #1376` to the commit message

michael-hoke replied:
Did this need to be added back in?

michael-hoke commented:
If you need to make wiki changes, do them and have them reviewed as a part of this PR. The checkbox is also meant to indicate to reviewers that there are documentation changes needed that they need to review.

JeremyYao commented:
> Why does this PR touch the default value for tdmlConfig.path?

If the user can't specify a custom pathing, then it should be reflected in how the pathing for the generated TDML is set.

michael-hoke commented:
> > Why does this PR touch the default value for tdmlConfig.path?
> 
> If the user can't specify a custom pathing, then it should be reflected in how the pathing for the generated TDML is set.

I'm not sure I follow. Are you saying that since we're not allowing the users to change the path, we should default to an empty string and then hardcode it right before we start the debugging session?

JeremyYao commented:
> > > Why does this PR touch the default value for tdmlConfig.path?
> > 
> > 
> > If the user can't specify a custom pathing, then it should be reflected in how the pathing for the generated TDML is set.
> 
> I'm not sure I follow. Are you saying that since we're not allowing the users to change the path, we should default to an empty string and then hardcode it right before we start the debugging session?

The default TDML config values are the starting initialization values for when we're creating new launch configurations. So, what this means is that '' is the default value.

michael-hoke commented:
> > > > Why does this PR touch the default value for tdmlConfig.path?
> > > 
> > > 
> > > If the user can't specify a custom pathing, then it should be reflected in how the pathing for the generated TDML is set.
> > 
> > 
> > I'm not sure I follow. Are you saying that since we're not allowing the users to change the path, we should default to an empty string and then hardcode it right before we start the debugging session?
> 
> The default TDML config values are the starting initialization values for when we're creating new launch configurations. So, what this means is that '' is the default value.

I do not think that '' should be the default value. What if we used `${AskForTDMLPath}` as the default value and added a block in debugger.ts:getDebugger to handle this for the two cases? I think this should also let us separate our PRs again.

JeremyYao commented on src/daffodilDebugger/debugger.ts:
```diff
@@ -58,12 +58,18 @@ async function getTDMLConfig(
-        'extension.dfdl-debug.getValidatedTDMLPath'
-      ))
+      config.tdmlConfig.path == '${AskForValidatedTDMLPath}'
+        ? await vscode.commands.executeCommand(
+            'extension.dfdl-debug.getValidatedTDMLPath'
+          )
+        : config.tdmlConfig.path
 
-    if (!config.tdmlConfig.path) return false
+    if (!config.tdmlConfig.path) {
+      vscode.window.showErrorMessage(
+        `Invalid path TDML Execute path ${config.tdmlConfig.path} was specified!`
+      )
+      return false
+    }
```
The check for undefined `tdmlConfig.path` is already defined via the 

```Typescript
    if (!config.tdmlConfig.path) {
      vscode.window.showErrorMessage(
        `Invalid path TDML Execute path ${config.tdmlConfig.path} was specified!`
      )
      return false
    }
``` 

block. The value `!undefined` evaluates to `true`. Also `!''`` returns true as well.

JeremyYao replied:
> If the user sends in an undefined or empty tdmlConfig.path, they will no longer be prompted for a path. I don't think that behavior should be regressed. If we see an undefined/empty path, I think we should prompt the user for the path instead of letting them know they made a mistake with their launch config and forcing them to fix it and restart.

Re-added via https://github.com/apache/daffodil-vscode/pull/1403/commits/bcb1c9bf52cd1e6ced778f38e65310366b7a5978.

michael-hoke commented on src/daffodilDebugger/debugger.ts:
```diff
@@ -58,12 +58,16 @@ async function getTDMLConfig(
-      config.tdmlConfig.path ||
-      (await vscode.commands.executeCommand(
-        'extension.dfdl-debug.getValidatedTDMLPath'
-      ))
+      config.tdmlConfig.path == '${AskForValidatedTDMLPath}' ||
+      !config.tdmlConfig.path
+        ? await vscode.commands.executeCommand(
+            'extension.dfdl-debug.getValidatedTDMLPath'
+          )
+        : config.tdmlConfig.path
```
This is a case where a ternary is harder to read than an if statement. The `config.tdmlConfig.path == '${AskForValidatedTDMLPath}'` is the main culprit here (and should be ===).

``` 
if (!config.tdmlConfig.path || config.tdmlConfig.path === '${AskForValidatedTDMLPath}')
{
  config.tdmlConfig.path = await vscode.commands.executeCommand('extension.dfdl-debug.getValidatedTDMLPath')
}
```

shanedell commented:
@JeremyYao Change

```txt
- [x] I have added documentation to the wiki for these changes 
```

to

```txt
- [x] I have added following documentation for these changes
```

and I will approved and merge after I rerun the enforce doc CI, which should pass after this change.

JeremyYao commented:
> @JeremyYao Change
> 
> ```
> - [x] I have added documentation to the wiki for these changes 
> ```
> 
> to
> 
> ```
> - [x] I have added following documentation for these changes
> ```
> 
> and I will approved and merge after I rerun the enforce doc CI, which should pass after this change.

Done

michael-hoke commented:
Looks ready to merge - created #1418 to track related bug

JeremyYao commented:
> Looks almost ready to merge - created #1418 to track related bug
> 
> Need to change references of ${AskForValidatedTDMLPath} to ${command:AskForValidatedTDMLPath}

Done in https://github.com/apache/daffodil-vscode/commit/fe97ae329938e99d463022daa381348d1bd74477
